File: CC++/sort/QuickSort.cpp
#include <iostream>
using namespace std;

int partition(int arr[], int low, int high) {
    int pivot = arr[high],
        left = low,
        right = high - 1;
    
    while (true) {
        while (left <= right && arr[left] > pivot) left++;
        while (left <= right && arr[right] < pivot) right--;
        if (left >= right) break;
        swap(arr[left], arr[right]);
        left++; right--;
    }

    swap(arr[left], arr[high]);
    return left;
}

void quickSort(int arr[], int low, int high) {
    if (low < high) {
        int pivot = partition(arr, low, high);

        quickSort(arr, low, pivot - 1);
        quickSort(arr, pivot + 1, high);
    }
}

int main() {
    int arr[] = {1, 2, 3, 4, 5, 6, 7, 8};
    quickSort(arr, 0, sizeof(arr)/sizeof(*arr) - 1);

    for (int i : arr)
        cout << i << " ";
}

File: Java/sort/QuickSort.java
class QuickSort {
    static void swap(int[] arr, int i, int j) {
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }

    static int partition(int[] arr, int low, int high) {
        int pivot = arr[high];
        int i = (low - 1);

        for (int j = low; j < high; j++)
            if (arr[j] < pivot) {
                i++;
                swap(arr, i, j);
            }

        swap(arr, i + 1, high);
        return i + 1;
    }

    static void quickSort(int[] arr, int low, int high) {
        if (low < high) {
            int pi = partition(arr, low, high);

            quickSort(arr, low, pi - 1);
            quickSort(arr, pi + 1, high);
        }
    }

    static void quickSort(int arr[]) {
        quickSort(arr, 0, arr.length - 1);
    }
    // Driver Code
    public static void main(String[] args) {
        int[] arr = { 10, 7, 8, 9, 1, 5, 6, 3, 8, 9, 34, 5 };

        quickSort(arr);
        
        for (int i : arr)
            System.out.print(i + " ");
    }
}